test/utils/image/csi-manifest.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This file is generated and managed by test/e2e/testing-manifests/storage-csi/update-hostpath.sh
+Do not edit
+*/
+
+package image
+
+const (
+	// Offset the CSI images so there is no collision
+	CSINone = iota + 500
+	Hostpathplugin
+	Csiexternalhealthmonitorcontroller
+	Csinodedriverregistrar
+	Livenessprobe
+	Csiattacher
+	Csiprovisioner
+	Csiresizer
+	Csisnapshotter
+)
+
+type TestCSIImagesStruct struct {
+	HostpathpluginImage                     string
+	CsiexternalhealthmonitorcontrollerImage string
+	CsinodedriverregistrarImage             string
+	LivenessprobeImage                      string
+	CsiattacherImage                        string
+	CsiprovisionerImage                     string
+	CsiresizerImage                         string
+	CsisnapshotterImage                     string
+}
+
+var TestCSIImages TestCSIImagesStruct
+
+func init() {
+	TestCSIImages = TestCSIImagesStruct{
+		GetE2EImage(Hostpathplugin),
+		GetE2EImage(Csiexternalhealthmonitorcontroller),
+		GetE2EImage(Csinodedriverregistrar),
+		GetE2EImage(Livenessprobe),
+		GetE2EImage(Csiattacher),
+		GetE2EImage(Csiprovisioner),
+		GetE2EImage(Csiresizer),
+		GetE2EImage(Csisnapshotter),
+	}
+}
+
+func initCSIImageConfigs(list RegistryList, configs map[int]Config) {
```

Comment:
Instead of generating this file with rather complex shell code, wouldn't it be simpler to just parse the embedded YAML files here and extract the image versions on-the-fly?

Reply:
I bet that would not only be easier to read and debug, but also less code.

Reply:
/hold

Please give me a bit more time to explore that approach.

Reply:
Please have a look at https://github.com/kubernetes/kubernetes/pull/108458

In contrast to this PR here it's done in pure Go and supports different image versions.

Reply:
That would mean that something in manifest.go would need to have a list of yaml files to parse. The "authoritative" list is deep in the storage e2es and I am afraid of circular deps.

https://github.com/kubernetes/kubernetes/blob/4737127b57d02051dfa9f5f263bcb5049516db1c/test/e2e/storage/drivers/csi.go#L165-L172

Of course, it's either possible to parse every embedded yaml file or have the list public somewhere else.

Reply:
We can iterate over all YAML files under test/e2e/testing-manifests without having to know how or by which tests those files are used. The resulting list of images might be a bit too broad, but that depends on which tests are going to run. I think it's okay to err on the side of caution and mirror anything that might be used.